open-company/open-company-api
# Minor API changes

Trello: https://trello.com/c/AWy33aAt

This PR fixes a hodgepodge of minor API issues:

* Defaults all section headlines to blanks
* Makes image and description read only properties of sections
* Remove PUT from company
* Makes PATCHed sections w/ no content get created as placeholders
* Disallow PATCHing made up sections into company
* normalize s/ as namespace alias for string and schema/ for prismatic schema
* Update icon set (also move it to S3 for easier updating w/o code changes)
* Ensures HATEOAS modify links are only present on individual section retrieval when they should be
* Ensures all the usage documented in the README works

To test:

* Look most closely at the test updates in the code changes. Ensure the tests are passing
* Observe the current new topic creation behavior in the client UI, this is enabled by these changes, does it seem good?
* Remove an existing topic that has some content, save it, then add the topic back, do you get the old content? This is enabled by these changes.
* Retrieve a section anonymously using cURL: `curl -i http://localhost:3000/companies/buffer/update` No update links anymore?
* Retrieve a section authorized, but back in time (with as-of). Easiest way to do this is to use the web UI on Buffer, it proactively caches the prior section revisions to the most recent. Look at the responses to those section retrievals with as-of in developer tools. No update links?
* Run through all the suggested code in the API README (start w/ a clean database). Does every line work now?

NOTE: The API is still incomplete as it relates to PATCHing sections. Right now, new sections can be PATCHed in blank (and they'll be placeholders, or the prior existing section), but they can't be PATCHed in with partial content (the rest of the content should come from the placeholder or prior section). There are some future-fact tests for this capability, but it doesn't exist yet. Primary reason being, I think we need to simplify the data model a bit and remove the sections being duplicated in the company document. It's just too complicated. No point implementing this full desired API, since our UI doesn't need it right now,  until we change that.
